include/jemalloc/internal/hpa.h

```diff
@@ -53,44 +27,46 @@ struct hpa_shard_s {
 	 * allocator, and so will use its edata_cache.
 	 */
 	edata_cache_small_t ecs;
-	hpa_t *hpa;
```

Comment:
Is it true that, this diff removed the central huge page allocator completely? If so what's the high level motivation behind it? My impression was that the hpa level contention didn't hurt much, and it potentially can enable flexible sharing? I also had hope to deal with "global" purging at the hpa level.

Reply:
Right; hpa_central could be deleted after this stack. A couple things pushed in this direction:
- hpa_central was actually the source of quite a bit of contention; it was by far the hottest of our locks on test binaries.
- Splitting a single hugepage across multiple suballocators means that there's no global view of it for the purposes of making purging / hugification decisions, which turns out to be important both for fragmentation avoidance and purging choices.

Practically this seems to be well-borne out; this design is performing far better than the previous one in canaries.

I think eventually we'll want the central allocator back, but only once we start trying to support large allocations (ones too big to serve out of the hpa_shard_t, or in case of hpa_shard_t exhaustion) and as the source of individual hpdata_ts for the hpa_shard_t.

Perhaps more intuitively, if thinking of it in terms of the earlier design: if we make the slab goal a whole hugepage, we can implement more efficient purging / hugification policies for free. But once we've done that (and, given that we don't actually try to serve large allocations yet), the hpa_central_t doesn't really add much (but still costs some complexity and lock contention).

Reply:
Sounds good. I think we are on the same page here -- 

> I think eventually we'll want the central allocator back, but only once we start trying to support large allocations